Can't generate nuget package from .NET 6 MacOS bindings project

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨 -->

<!--
Please fill in the *entire* template below.
The template is designed to avoid unnecessary delays to confirm and fix reported issues. Issues with missing information generally end up with a `need-info` label and take significantly more time to fix. Please help us help you!
For more information see https://github.com/xamarin/xamarin-macios/wiki/Submitting-Bugs-&-Suggestions
For support requests use https://www.xamarin.com/support
-->
### Steps to Reproduce

1. Create .NET 6 macos bindings project
2. Add a framework eg. [MetalANGLE 0.0.7](https://github.com/kakashidinho/metalangle/releases/download/gles3-0.0.7/MetalANGLE.framework.mac.zip) (what I'm using)
3. Build the project with `dotnet build --configuration Release -p:GeneratePackageOnBuild=true -p:IncludeSymbols=true -p:SymbolPackageFormat=snupkg -p:DebugSymbols=true -p:DebugType=portable` to generate a nuget package

### Expected Behavior
The project builds successfully and generates a nuget package.

### Actual Behavior
The build fails with the following error:
```
/usr/local/share/dotnet/sdk/6.0.100-rc.2.21425.12/Sdks/NuGet.Build.Tasks.Pack/build/NuGet.Build.Tasks.Pack.targets(221,5): error NU5026: The file '/Users/Hamish/Projects/Tests/TestBindingsProject2/TestBindingsProject/bin/Release/net6.0-macos/Native.TestBindingsProject.manifest' to be packed was not found on disk. [/Users/Hamish/Projects/Tests/TestBindingsProject2/TestBindingsProject/TestBindingsProject.csproj]

Build FAILED.

/usr/local/share/dotnet/sdk/6.0.100-rc.2.21425.12/Sdks/NuGet.Build.Tasks.Pack/build/NuGet.Build.Tasks.Pack.targets(221,5): error NU5026: The file '/Users/Hamish/Projects/Tests/TestBindingsProject2/TestBindingsProject/bin/Release/net6.0-macos/Native.TestBindingsProject.manifest' to be packed was not found on disk. [/Users/Hamish/Projects/Tests/TestBindingsProject2/TestBindingsProject/TestBindingsProject.csproj]
```

### Environment

<!--
1.
Visual Studio: Help > About Microsoft Visual Studio > Copy Info [button]
Visual Studio for Mac: Visual Studio > About Visual Studio > Show Details > Copy Information [button]
2. Paste into the code block below (between ```)
-->

```
Visual Studio Community 2022 for Mac Preview
Version 17.0 Preview (17.0 build 2706)
Installation UUID: 93d39ea7-8431-4eae-9b2f-0b0bd10e30da
	GTK+ 2.24.23

Mono Framework MDK
Runtime:
	Mono 6.12.0 (2020-02/eb4c3116ebf) (64-bit)
	Package version: 612000147

Roslyn (Language Service)
3.10.0-3.21226.12+c153357eb74f4b5c8fc905b7ae42168b76515531

NuGet
Version: 6.0.0.129

.NET SDK
SDK: /usr/local/share/dotnet/sdk/6.0.100-rc.2.21425.12/Sdks
SDK Versions:
	6.0.100-rc.2.21425.12
	6.0.100-rc.1.21417.3
	6.0.100-rc.1.21403.66
	6.0.100-preview.7.21379.14
	5.0.400
	3.1.412
	3.0.103
	2.2.207
	2.1.817
MSBuild SDKs: /Applications/Visual Studio (Preview).app/Contents/MonoBundle/MSBuild/Current/bin/Sdks

.NET Core Runtime
Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	6.0.0-rc.2.21423.6
	6.0.0-rc.1.21415.6
	6.0.0-rc.1.21401.3
	5.0.9
	3.1.18
	3.0.3
	2.2.8
	2.1.29

.NET 5.0 SDK
SDK: 5.0.400

.NET Core 3.1 SDK
SDK: 3.1.412

Xamarin.Profiler
Version: 1.6.15.68
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Updater
Version: 11

Xamarin Designer
Version: 17.0.0.58
Hash: 927d548a1
Branch: remotes/origin/d17-0
Build date: 2021-08-27 11:27:00 UTC

Apple Developer Tools
Xcode 13.0 (19223.2)
Build 13A5212g

Xamarin.Mac
Version: 7.14.0.24 (Visual Studio Community)
Hash: c4b89cddb
Branch: d16-10
Build date: 2021-06-15 22:03:00-0400

Xamarin.iOS
Version: 14.20.0.24 (Visual Studio Community)
Hash: c4b89cddb
Branch: d16-10
Build date: 2021-06-15 22:03:01-0400

Xamarin.Android
Version: 11.3.99.54 (Visual Studio Community)
Commit: xamarin-android/main/0e5e06f
Android SDK: /Users/Hamish/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		6.0 (API level 23)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 31.0.2
SDK Build Tools Version: 30.0.2

Build Information: 
Mono: c633fe9
Java.Interop: xamarin/java.interop/main@a5ed891
ProGuard: Guardsquare/proguard/v7.0.1@912d149
SQLite: xamarin/sqlite/3.35.4@85460d3
Xamarin.Android Tools: xamarin/xamarin-android-tools/main@683f375

Microsoft OpenJDK for Mobile
Java SDK: /Users/Hamish/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Android SDK Manager
Version: 17.0.0.3
Hash: f06a85b
Branch: remotes/origin/HEAD
Build date: 2021-08-27 11:27:02 UTC

Android Device Manager
Version: 17.0.0.2
Hash: d8af29f
Branch: remotes/origin/main
Build date: 2021-08-27 11:27:02 UTC

Build Information
Release ID: 1700002706
Git revision: ba899d0dec8c3161225334edfcc5569ebfae78cc
Build date: 2021-08-27 07:26:10-04
Build branch: release-17.0

Operating System
Mac OS X 11.5.2
Darwin 20.6.0 Darwin Kernel Version 20.6.0
    Wed Jun 23 00:26:31 PDT 2021
    root:xnu-7195.141.2~5/RELEASE_X86_64 x86_64
```

### Build Logs

<!--
1. Place cursor below this comment block.
2. Attach build log or link to gist (https://gist.github.com/) of the log.
3. Logs can be found in the Build Output tab of the Errors pad in Visual Studio
To get full build logs, set the log verbosity to diagnostic at the following locations:
- On Visual Studio for Mac: Preferences > Projects > Build
- On Visual Studio for Windows: Tools > Options > Projects and Solutions > Build and Run
On Visual Studio for Windows you also want to add `-v -v -v -v` to the mtouch additional arguments by right-clicking the project in the solution explorer and selecting `Properties`.
-->
[msbuild.log.zip](https://github.com/xamarin/xamarin-macios/files/7115104/msbuild.log.zip)


### Example Project (If Possible)

<!--
1. Place cursor below this comment block.
2. Drag and drop the compressed project or files needed to reproduce.
-->

<!--
Switch to the "Preview" tab to ensure your issue renders correctly.
-->
[TestBindingsProject2.zip](https://github.com/xamarin/xamarin-macios/files/7115120/TestBindingsProject2.zip)
Note, before building this project, you need to first extract the zip at TestBindingsProject/Natives/MetalANGLE.framework.mac.zip first using Archive Utility (other unarchivers may also work) - it should create a folder called MetalANGLE.framework in the Natives folder (same structure as the framework for the project in #12630).

--- a/msbuild/Xamarin.Shared/Xamarin.Shared.targets
+++ b/msbuild/Xamarin.Shared/Xamarin.Shared.targets
@@ -194,6 +194,68 @@ Copyright (C) 2018 Microsoft. All rights reserved.
 		</Touch>
 	</Target>
 
+	<!--
+		Binding projects may include NativeReference items, which makes the 'pack' target want to include a Native.$(AssemblyName).manifest into the NuGet,
+		which obviously fails because we don't create such a file. So let's remove that file.
+
+		Ref: https://github.com/dotnet/msbuild/blob/3a1e456fe227f3e2b190b434578844c31e8bcb4a/src/Tasks/Microsoft.Common.CurrentVersion.targets#L6111-L6118
+		Ref: https://github.com/dotnet/msbuild/issues/4584
+	-->
+	<PropertyGroup>
+		<GenerateNuspecDependsOn>
+			$(GenerateNuspecDependsOn);
+			_RemoveNativeManifestFromPack;
+		</GenerateNuspecDependsOn>
+	</PropertyGroup>
+	<Target Name="_RemoveNativeManifestFromPack">
+		<ItemGroup>
+			<_BuildOutputInPackage Remove="@(_BuildOutputInPackage)" Condition="'%(Filename)%(Extension)' == '$(_DeploymentTargetApplicationManifestFileName)'" />
+		</ItemGroup>
+	</Target>
+
+	<!--
+		Add any binding resource packages to the nupkg
+
+		Ref: https://github.com/dotnet/sdk/issues/14042#issuecomment-716868311
+		Ref: https://github.com/NuGet/Home/issues/10063#issuecomment-713083004
+		Ref: https://github.com/xamarin/xamarin-android/blob/681887ebdbd192ce7ce1cd02221d4939599ba762/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AndroidLibraries.targets#L86-L99
+	-->
+	<PropertyGroup>
+		<TargetsForTfmSpecificContentInPackage>$(TargetsForTfmSpecificContentInPackage);_IncludeBindingResourcesInNuGetPackage</TargetsForTfmSpecificContentInPackage>
+	</PropertyGroup>
+	<Target Name="_IncludeBindingResourcesInNuGetPackage"
+		Condition="'$(IsMacEnabled)' == 'true' And '$(IncludeBuildOutput)' != 'false'"
+	>
+		<PropertyGroup>
+			<_HasOldStyleBindingItems Condition="@(ObjcBindingNativeLibrary->Count()) > 0">true</_HasOldStyleBindingItems>
+		</PropertyGroup>
+		<Error Condition="'$(_HasOldStyleBindingItems)' == 'true'" Text="Creating a NuGet package is not supported for projects that have ObjcBindingNativeLibrary items. Migrate to use NativeReference items instead." />
+
+		<!-- Figure out where to place the binding resources (see references above for more info) -->
+		<GetNuGetShortFolderName
+			TargetFrameworkMoniker="$(TargetFrameworkMoniker)"
+			TargetPlatformMoniker="$(TargetPlatformMoniker)">
+			<Output TaskParameter="NuGetShortFolderName" PropertyName="_NuGetShortFolderName" />
+		</GetNuGetShortFolderName>
+
+		<!-- The binding project may either produce a 'MyBindingProject.resources' directory, or a compressed 'MyBindingProject.resources.zip' version -->
+		<PropertyGroup>
+			<_AppleBindingResourceBasePath>$(TargetDir)$(TargetName).resources</_AppleBindingResourceBasePath>
+		</PropertyGroup>
+
+		<ItemGroup>
+			<!-- Add all the files that might exist in any binding resource packages -->
+			<_AppleBindingResource Include="$(_AppleBindingResourceBasePath)\**\*" PackagePath="lib\$(_NuGetShortFolderName)\$(TargetName).resources" />
+			<!-- Add any compressed files that might exist as well -->
+			<_AppleBindingResource Include="$(_AppleBindingResourceBasePath).zip" PackagePath="lib\$(_NuGetShortFolderName)" Condition="Exists('$(_AppleBindingResourceBasePath).zip')" />
+		</ItemGroup>
+
+		<!-- Add what we found to the NuGet -->
+		<ItemGroup>
+			<TfmSpecificPackageFile Include="@(_AppleBindingResource)" />
+		</ItemGroup>
+	</Target>
+
 	<!-- Cleaning via FileWrites leaves empty framework directories on disk, so nuke via RemoveDir -->
 	<PropertyGroup>
 		<CleanDependsOn>

--- a/msbuild/Xamarin.Shared/Xamarin.Shared.targets
+++ b/msbuild/Xamarin.Shared/Xamarin.Shared.targets
@@ -156,7 +156,7 @@ Copyright (C) 2018 Microsoft. All rights reserved.
 	</Target>
 	
 	<PropertyGroup>
-		<BindingResourcePath>$(ProjectDir)$(OutputPath)$(AssemblyName).resources</BindingResourcePath>
+		<BindingResourcePath>$(OutputPath)$(AssemblyName).resources</BindingResourcePath>
 	</PropertyGroup>
 
 	<Target Name="_CreateBindingResourcePackage" Condition="'$(DesignTimeBuild)' != 'true'"

--- a/tests/common/DotNet.cs
+++ b/tests/common/DotNet.cs
@@ -24,6 +24,18 @@ namespace Xamarin.Tests {
 			}
 		}
 
+		public static ExecutionResult AssertPack (string project, Dictionary<string, string> properties = null)
+		{
+			return Execute ("pack", project, properties, true);
+		}
+
+		public static ExecutionResult AssertPackFailure (string project, Dictionary<string, string> properties = null)
+		{
+			var rv = Execute ("pack", project, properties, false);
+			Assert.AreNotEqual (0, rv.ExitCode, "Unexpected success");
+			return rv;
+		}
+
 		public static ExecutionResult AssertPublish (string project, Dictionary<string, string> properties = null)
 		{
 			return Execute ("publish", project, properties, true);
@@ -75,6 +87,7 @@ namespace Xamarin.Tests {
 			switch (verb) {
 			case "clean":
 			case "build":
+			case "pack":
 			case "publish":
 				var args = new List<string> ();
 				args.Add (verb);

--- a/tests/dotnet/MyClassLibrary/MyClassLibrary.csproj
+++ b/tests/dotnet/MyClassLibrary/MyClassLibrary.csproj
@@ -1,6 +1,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<Project Sdk="Microsoft.NET.Sdk">
-  <PropertyGroup>
-    <TargetFramework>net6.0-ios</TargetFramework>
-  </PropertyGroup>
-</Project>

--- a/tests/dotnet/UnitTests/ApplePlatformExtensions.cs
+++ b/tests/dotnet/UnitTests/ApplePlatformExtensions.cs
@@ -1,0 +1,22 @@
+using System;
+
+namespace Xamarin.Utils {
+	public static class ApplePlatformExtensionsWithVersions {
+		public static string ToFrameworkWithDefaultVersion (this ApplePlatform @this)
+		{
+			var netVersion = "net6.0";
+			switch (@this) {
+			case ApplePlatform.iOS:
+				return netVersion + "-ios" + SdkVersions.iOS;
+			case ApplePlatform.MacOSX:
+				return netVersion + "-macos" + SdkVersions.OSX;
+			case ApplePlatform.TVOS:
+				return netVersion + "-tvos" + SdkVersions.TVOS;
+			case ApplePlatform.MacCatalyst:
+				return netVersion + "-maccatalyst" + SdkVersions.MacCatalyst;
+			default:
+				return "Unknown";
+			}
+		}
+	}
+}

--- a/tests/dotnet/UnitTests/DotNetUnitTests.csproj
+++ b/tests/dotnet/UnitTests/DotNetUnitTests.csproj
@@ -48,6 +48,9 @@
     <Compile Include="..\..\..\external\Xamarin.MacDev\Xamarin.MacDev\PListObject.cs">
       <Link>external\PListObject.cs</Link>
     </Compile>
+    <Compile Include="..\..\..\tools\common\SdkVersions.cs">
+      <Link>external\SdkVersions.cs</Link>
+    </Compile>
   </ItemGroup>
   <ItemGroup>
     <Folder Include="external\" />

--- a/tests/dotnet/UnitTests/PackTest.cs
+++ b/tests/dotnet/UnitTests/PackTest.cs
@@ -1,0 +1,175 @@
+#nullable enable
+
+using System;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+
+using NUnit.Framework;
+
+using Xamarin.Utils;
+using Xamarin.MacDev;
+
+namespace Xamarin.Tests {
+	public class PackTest : TestBaseClass {
+
+
+		[Test]
+		[TestCase (ApplePlatform.iOS)]
+		[TestCase (ApplePlatform.MacCatalyst)]
+		[TestCase (ApplePlatform.TVOS)]
+		[TestCase (ApplePlatform.MacOSX)]
+		public void BindingProject (ApplePlatform platform)
+		{
+			var project = "bindings-test";
+			Configuration.IgnoreIfIgnoredPlatform (platform);
+
+			var project_path = Path.Combine (Configuration.RootPath, "tests", project, "dotnet", platform.AsString (), $"{project}.csproj");
+			Clean (project_path);
+			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName (project_path));
+
+			var tmpdir = Cache.CreateTemporaryDirectory ();
+			var outputPath = Path.Combine (tmpdir, "OutputPath");
+			var intermediateOutputPath = Path.Combine (tmpdir, "IntermediateOutputPath");
+			var properties = GetDefaultProperties ();
+			properties ["OutputPath"] = outputPath + Path.DirectorySeparatorChar;
+			properties ["IntermediateOutputPath"] = intermediateOutputPath + Path.DirectorySeparatorChar;
+
+			var rv =DotNet.AssertPackFailure (project_path, properties);
+			var errors = BinLog.GetBuildLogErrors (rv.BinLogPath).ToArray ();
+			Assert.AreEqual (1, errors.Length, "Error count");
+			Assert.AreEqual ($"Creating a NuGet package is not supported for projects that have ObjcBindingNativeLibrary items. Migrate to use NativeReference items instead.", errors [0].Message, "Error message");
+		}
+
+		[Test]
+		[TestCase (ApplePlatform.iOS, true)]
+		[TestCase (ApplePlatform.iOS, false)]
+		[TestCase (ApplePlatform.MacCatalyst, true)]
+		[TestCase (ApplePlatform.MacCatalyst, false)]
+		[TestCase (ApplePlatform.TVOS, true)]
+		[TestCase (ApplePlatform.TVOS, false)]
+		[TestCase (ApplePlatform.MacOSX, true)]
+		[TestCase (ApplePlatform.MacOSX, false)]
+		public void BindingFrameworksProject (ApplePlatform platform, bool noBindingEmbedding)
+		{
+			var project = "bindings-framework-test";
+			Configuration.IgnoreIfIgnoredPlatform (platform);
+
+			var project_path = Path.Combine (Configuration.RootPath, "tests", project, "dotnet", platform.AsString (), $"{project}.csproj");
+			Clean (project_path);
+			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName (project_path));
+
+			var tmpdir = Cache.CreateTemporaryDirectory ();
+			var outputPath = Path.Combine (tmpdir, "OutputPath");
+			var intermediateOutputPath = Path.Combine (tmpdir, "IntermediateOutputPath");
+			var properties = GetDefaultProperties ();
+			properties ["OutputPath"] = outputPath + Path.DirectorySeparatorChar;
+			properties ["IntermediateOutputPath"] = intermediateOutputPath + Path.DirectorySeparatorChar;
+			properties ["NoBindingEmbedding"] = noBindingEmbedding ? "true" : "false";
+
+			DotNet.AssertPack (project_path, properties);
+
+			var nupkg = Path.Combine (outputPath, project + ".1.0.0.nupkg");
+			Assert.That (nupkg, Does.Exist, "nupkg existence");
+
+			var archive = ZipFile.OpenRead (nupkg);
+			var files = archive.Entries.Select (v => v.FullName).ToHashSet ();
+			var hasSymlinks = noBindingEmbedding && (platform == ApplePlatform.MacCatalyst || platform == ApplePlatform.MacOSX);
+			if (noBindingEmbedding) {
+				Assert.That (archive.Entries.Count, Is.EqualTo (hasSymlinks ? 6 : 10), $"nupkg file count - {nupkg}");
+			} else {
+				Assert.That (archive.Entries.Count, Is.EqualTo (5), $"nupkg file count - {nupkg}");
+			}
+			Assert.That (files, Does.Contain (project + ".nuspec"), "nuspec");
+			Assert.That (files, Does.Contain ("_rels/.rels"), ".rels");
+			Assert.That (files, Does.Contain ("[Content_Types].xml"), "[Content_Types].xml");
+			Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.dll"), $"{project}.dll");
+			Assert.That (files, Has.Some.Matches<string> (v => v.StartsWith ("package/services/metadata/core-properties/", StringComparison.Ordinal) && v.EndsWith (".psmdcp", StringComparison.Ordinal)), "psmdcp");
+			if (noBindingEmbedding) {
+				if (hasSymlinks) {
+					Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.resources.zip"), $"{project}.resources.zip");
+				} else {
+					Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.resources/XStaticArTest.framework/XStaticArTest"), $"XStaticArTest.framework/XStaticArTest");
+					Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.resources/XStaticObjectTest.framework/XStaticObjectTest"), $"XStaticObjectTest.framework/XStaticObjectTest");
+					Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.resources/XTest.framework/XTest"), $"XTest.framework/XTest");
+					Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.resources/XTest.framework/Info.plist"), $"XTest.framework/Info.plist");
+					Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.resources/manifest"), $"manifest");
+				}
+			}
+		}
+
+		[Test]
+		[TestCase (ApplePlatform.iOS, true)]
+		[TestCase (ApplePlatform.iOS, false)]
+		[TestCase (ApplePlatform.MacCatalyst, true)]
+		[TestCase (ApplePlatform.MacCatalyst, false)]
+		[TestCase (ApplePlatform.TVOS, true)]
+		[TestCase (ApplePlatform.TVOS, false)]
+		[TestCase (ApplePlatform.MacOSX, true)]
+		[TestCase (ApplePlatform.MacOSX, false)]
+		public void BindingXcFrameworksProject (ApplePlatform platform, bool noBindingEmbedding)
+		{
+			var project = "bindings-xcframework-test";
+			Configuration.IgnoreIfIgnoredPlatform (platform);
+
+			var project_path = Path.Combine (Configuration.RootPath, "tests", project, "dotnet", platform.AsString (), $"{project}.csproj");
+			Clean (project_path);
+			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName (project_path));
+
+			var tmpdir = Cache.CreateTemporaryDirectory ();
+			var outputPath = Path.Combine (tmpdir, "OutputPath");
+			var intermediateOutputPath = Path.Combine (tmpdir, "IntermediateOutputPath");
+			var properties = GetDefaultProperties ();
+			properties ["OutputPath"] = outputPath + Path.DirectorySeparatorChar;
+			properties ["IntermediateOutputPath"] = intermediateOutputPath + Path.DirectorySeparatorChar;
+			properties ["NoBindingEmbedding"] = noBindingEmbedding ? "true" : "false";
+			properties ["AssemblyName"] = project;
+
+			DotNet.AssertPack (project_path, properties);
+
+			var nupkg = Path.Combine (outputPath, project + ".1.0.0.nupkg");
+			Assert.That (nupkg, Does.Exist, "nupkg existence");
+
+			var archive = ZipFile.OpenRead (nupkg);
+			var files = archive.Entries.Select (v => v.FullName).ToHashSet ();
+			Assert.That (archive.Entries.Count, Is.EqualTo (noBindingEmbedding ? 6 : 5), $"nupkg file count - {nupkg}");
+			Assert.That (files, Does.Contain (project + ".nuspec"), "nuspec");
+			Assert.That (files, Does.Contain ("_rels/.rels"), ".rels");
+			Assert.That (files, Does.Contain ("[Content_Types].xml"), "[Content_Types].xml");
+			Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.dll"), $"{project}.dll");
+			Assert.That (files, Has.Some.Matches<string> (v => v.StartsWith ("package/services/metadata/core-properties/", StringComparison.Ordinal) && v.EndsWith (".psmdcp", StringComparison.Ordinal)), "psmdcp");
+			if (noBindingEmbedding) {
+				Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.resources.zip"), $"{project}.resources.zip");
+			}
+		}
+
+		[Test]
+		[TestCase (ApplePlatform.iOS)]
+		[TestCase (ApplePlatform.MacCatalyst)]
+		[TestCase (ApplePlatform.TVOS)]
+		[TestCase (ApplePlatform.MacOSX)]
+		public void LibraryProject (ApplePlatform platform)
+		{
+			var project = "MyClassLibrary";
+			Configuration.IgnoreIfIgnoredPlatform (platform);
+
+			var project_path = GetProjectPath (project, runtimeIdentifiers: string.Empty, platform: platform, out var appPath);
+			Clean (project_path);
+			var properties = GetDefaultProperties ();
+
+			DotNet.AssertPack (project_path, properties);
+
+			var nupkg = Path.Combine (Path.GetDirectoryName (project_path)!, "bin", "Debug", project + ".1.0.0.nupkg");
+			Assert.That (nupkg, Does.Exist, "nupkg existence");
+
+			var archive = ZipFile.OpenRead (nupkg);
+			var files = archive.Entries.Select (v => v.FullName).ToHashSet ();
+			Assert.That (archive.Entries.Count, Is.EqualTo (5), "nupkg file count");
+			Assert.That (files, Does.Contain (project + ".nuspec"), "nuspec");
+			Assert.That (files, Does.Contain ("_rels/.rels"), ".rels");
+			Assert.That (files, Does.Contain ("[Content_Types].xml"), "[Content_Types].xml");
+			Assert.That (files, Does.Contain ($"lib/{platform.ToFrameworkWithDefaultVersion ()}/{project}.dll"), $"{project}.dll");
+			Assert.That (files, Has.Some.Matches<string> (v => v.StartsWith ("package/services/metadata/core-properties/", StringComparison.Ordinal) && v.EndsWith (".psmdcp", StringComparison.Ordinal)), "psmdcp");
+		}
+	}
+}

